libtransmission/peer-mgr.h

```diff
@@ -316,17 +351,100 @@ class tr_peer_info
+            //   | T               | T             | T                    | F                  | T       |
+            //   | T               | T             | F                    | F                  | ?       |
+            //   | T               | T             | ?                    | T                  | T       |
+            //   | T               | T             | ?                    | F                  | T       |
+            //   | T               | F             | T                    | T                  | T       |
+            //   | T               | F             | T                    | F                  | T       |
+            //   | T               | F             | F                    | F                  | ?       |
+            //   | T               | F             | ?                    | T                  | T       |
+            //   | T               | F             | ?                    | F                  | T       |
+            //   | F               | F             | T                    | T                  | ?       |
+            //   | F               | F             | T                    | F                  | ?       |
+            //   | F               | F             | F                    | F                  | F       |
+            //   | F               | F             | ?                    | T                  | ?       |
+            //   | F               | F             | ?                    | F                  | F       |
+            //   | ?               | T             | T                    | T                  | T       |
+            //   | ?               | T             | T                    | F                  | T       |
+            //   | ?               | T             | F                    | F                  | ?       |
+            //   | ?               | T             | ?                    | T                  | ?       |
+            //   | ?               | T             | ?                    | F                  | ?       |
+            //   | ?               | F             | T                    | T                  | T       |
+            //   | ?               | F             | T                    | F                  | T       |
+            //   | ?               | F             | F                    | F                  | F       |
+            //   | ?               | F             | ?                    | T                  | ?       |
+            //   | ?               | F             | ?                    | F                  | ?       |
+            //   | N/A             | N/A           | F                    | T                  | Invalid |
+            //   | F               | T             | N/A                  | N/A                | Invalid |
+            //   +-----------------+---------------+----------------------+--------------------+---------+
```

Comment:
:exploding_head: 

I don't have a better suggestion and this is better than the existing code, but this complexity is a code smell that makes me wonder how we could ever triage/debug issues if any of the code is wrong.

Not a blocker + much of this complexity is coming from choices made _before_ this PR, but we should think about whether there's a way to make this better

Reply:
Agreed, this code is smelly.